### PR TITLE
Add --noprivate flag to describegraph rpc to filter out private channels

### LINKED ISSUE #1037: private channel in describegraph
### Background

channels that are private to the node are shown in `describegraph` .
can we add the `private` flag to describegraph so they are easier to filter out on visualizers etc?

### PR DESCRIPTION
**Pull request by Bluetegu**
Add --noprivate flag to describegraph rpc to filter out private channels per #1037 

Initially I named the flag `private` as suggested in the issue, but setting private to filter private seems wrong to me, and I didn't want to change the behaviour of describegraph that returns private channels by default, so I changed the flag name to `noprivate`

Note: this additional parameter needs to be added to the describegraph cache issue #1232 

Testing done on simnet:
Public channel between Alice and Bob
Private channel between Bob and Charlie

On Charlie's side, run describegraph and describegraph --noprivate and compare outputs. Only difference is that when --noprivate flag is on, the private channel was not included in the output.

Fixes #1037.

### CODE REVIEW

**Comment by halseth** [comment]
I would rather make private channels not appear by default, as that might make users posting their private channels by mistake. Instead `--private` should be provided _to include_ private channels.

**Comment by Bluetegu** [comment]
@halseth, I agree this makes more sense. Switched flag from 'noprivate' to 'private'.

File: cmd/lncli/commands.go [hunk 1]
```diff
@@ -1985,7 +1989,18 @@ func describeGraph(ctx *cli.Context) error {
...
+		private bool
+	)
+
+	switch {
+	case ctx.IsSet("private"):
+		private = true
+	}
+
+	req := &lnrpc.ChannelGraphRequest{
+		Private: bool(private),
```
Comment (wpaulino) [inline cmd/lncli/commands.go]: As it's a boolean value we're attempting to retrieve, we can simply do `Private: ctx.Bool("private"),` and remove the code above.

File: rpcserver.go [hunk 2]
```diff
@@ -2709,8 +2710,13 @@ func (r *rpcServer) DescribeGraph(ctx context.Context,
 	err = graph.ForEachChannel(func(edgeInfo *channeldb.ChannelEdgeInfo,
 		c1, c2 *channeldb.ChannelEdgePolicy) error {
 
-		edge := marshalDbEdge(edgeInfo, c1, c2)
-		resp.Edges = append(resp.Edges, edge)
+		// A private channel does not need an authentication proof, while an
```
Comment (wpaulino) [inline rpcserver.go]: nit: comment greater than 80 char limit.

File: rpcserver.go [hunk 3]
```diff
@@ -2709,8 +2710,13 @@ func (r *rpcServer) DescribeGraph(ctx context.Context,
...
+		// advertized channel must have one.
+		// Do not include private channels if unless specifically required.
+		if private || edgeInfo.AuthProof != nil {
+			edge := marshalDbEdge(edgeInfo, c1, c2)
+			resp.Edges = append(resp.Edges, edge)
+		}
```
Comment (wpaulino) [inline rpcserver.go]: nit: add newline after closing brace.

**Comment by Bluetegu** [comment]
@wpaulino , I made the changes per your review comments. Let me know if further changes need to be made. Thanks!

**Comment by halseth** [comment]
Commits that revert/changes a previous commit should be squashed together :)

Also needs a rebase and a proto-regen!

**Comment by Bluetegu** [comment]
@halseth 
I squashed the commits and rebased; my first squash ever, its a learning experience :-). Hope its ok now.

I've redone the test above to verify correct operation of describegraph with and without the private flag.

I try to make sure to do a proto gen upon each commit. I'm using the versions for the generator specified for lnd. It does though introduce a small number of changes in other places in code in rpc.pb.go and rpc.pb.gw.go. I patch those back, using the diff of running the proto gen on the master branch. Most of the changes seem trivial, i.e. one liner function instead of 3:

-type isOpenStatusUpdate_Update interface{ isOpenStatusUpdate_Update() }
+type isOpenStatusUpdate_Update interface {
+       isOpenStatusUpdate_Update()
+}

But also:

 func RegisterLightningHandler(ctx context.Context, mux *runtime.ServeMux, conn *grpc.ClientConn) error {
-       client := NewLightningClient(conn)
+       return RegisterLightningHandlerClient(ctx, mux, NewLightningClient(conn))
+}

I double checked the versions on golang/protobuf and grpc-ecosystem/grpc-gateway. 
And my protobuf version seems fine too:
ubuntu@btc:~$ protoc --version
libprotoc 3.4.0

I'm not sure what is causing this. I don't think this has consequences on this PR, but it would be nice to know how to fix it such that when I regen proto on master it will not generate any changes at all, as it should.

**Comment by Bluetegu** [comment]
@halseth , sorry, just realized that I'm pulling from my master, so indeed I need to rebase... 
I'll submit an updated pull request asap.

**Comment by Bluetegu** [comment]
@halseth, I rebased, re-gen proto, didn't do any weird patches this time, did the test mentioned above again to verify that it's working. Hope its now ready. Sorry for the noob-ness.

File: cmd/lncli/commands.go [hunk 4]
```diff
@@ -2306,6 +2306,10 @@ var describeGraphCommand = cli.Command{
 			Name:  "render",
 			Usage: "If set, then an image of graph will be generated and displayed. The generated image is stored within the current directory with a file name of 'graph.svg'",
 		},
+		cli.BoolFlag{
+			Name:  "private",
```
Comment (halseth) [inline cmd/lncli/commands.go]: Awesome, look like you were able to squash 👍 

Ideally, you would make one commit for all the changes in `lnrpc/`, one for changes in `cmd/` and one commit for the changes to `rpcserver.go`. Not a blocker though!

**Comment by Bluetegu** [comment]
@halseth, I rebased again, as master moved forward and a conflict was detected in rpc.bp.go. I rerun gen_proto, rebuild, go test, and run the specific tests above. 
But I see that this update dismissed you ok review. 
Was I wrong to rebase? What is the process here, should you re-ok?
Thanks!

**Comment by halseth** [comment]
No, that was correct! 👍  Will re-approve when we have been able to get Travis to run.

**Comment by Bluetegu** [comment]
Thanks! Travis support un-flagged my account, but pull requests that were submitted before the unflagging will need to be re-triggered manually or by an updated pull request :-(

**Comment by halseth** [comment]
@Bluetegu I think you should be able to trigger a build by either rebasing, or altering your commits, then force pushing :)

**Comment by Bluetegu** [comment]
@halseth I need some help here :-)

The code I added assumes that non-private channels, or more accurately their edgeInfo, always have a non-nill authProof. This works when I tested with lncli. However this check/assumption fails for non-private channels in the integration tests. I am not sure why though.

Looking at other code in rpcserver.go that checks for private channels, e.g. like ListChannels, the code checks whether the channel is private by comparing the channel flags against lnwire.FFAnnounceChannel

What I'm thinking of doing, is instead of checking if authProof is empty or not, add a function that gets all private channels, and somehow check whether the edgeInfo being iterated upon is private or not by checking whether it is/reflects one of those channels. 

Is it the right direction to go?

Thanks!

**Comment by halseth** [comment]
Non-private channel **should** have an authproof after the funding transaction has received 6 confirmations. However there can be cases where this doesn't get created, if one of the channel parties goes offline for instance before those 6 confirmations. In such cases the channel won't be active though, so it shouldn't be very interesting.

Using `FFAnnounceChannel` to filter out private channels is a possibility, but note that we only have these flags for our own channels. So in that case we would need to determine which channels we own, and query the database for the flags.

What integration test is failing because of this?

**Comment by Bluetegu** [comment]
@halseth The following tests fail:

'open_channel_reorg_test' fails at:
*errors.errorString expected to find one edge in the graph, found 0 
                /home/ubuntu/go/src/github.com/lightningnetwork/lnd/lnd_test.go:1166 (0xb6b4f9) 

'channel_force_closure' fails at: 
                *errors.errorString expected force closed channel to have 6 pending htlcs, found 0 instead
                /home/ubuntu/go/src/github.com/lightningnetwork/lnd/lnd_test.go:1615 (0xb69807)

'private_channels' fails at:
        lnd_test.go:74: Failed: (private channels): exited with error: 
                *errors.errorString expected Alice to know 4 edges, had 3
                /home/ubuntu/go/src/github.com/lightningnetwork/lnd/lnd_test.go:3876 (0xb7216a)

'test_multi-hop_htlc_local_chain_claim' fails too. I stopped there...

Regarding 'using FFAnnounceChannel to filter out private channels is a possibility, but note that we only have these flags for our own channels.', yes, but per my understanding we are not aware of private channels (edgeInfo) that are not ours, right? Otherwise, if we got those through gossip they are public. 

Thanks!

**Comment by halseth** [comment]
@Bluetegu I don't think your changes should affect the first two tests you mentioned, so that is probably a different bug... I think I have PRs up that fixes both these two flakes.

For the test `private_channels`, try to determine if it is a flake (fails just sometimes) or if it can be reproed constantly.

> Regarding 'using FFAnnounceChannel to filter out private channels is a possibility, but note that we only have these flags for our own channels.', yes, but per my understanding we are not aware of private channels (edgeInfo) that are not ours, right? Otherwise, if we got those through gossip they are public.

You are absolutely right, `edgeInfo`s for private channels won't propagate in the network :) What I'm saying is that `ForEachChannel` will return all channels, whether they are ours or not, and to check the `ChannelFlags` we must query the `channeldb` (instead of just the graph). Using the `authProof` directly lets us avoid that.

**Comment by Bluetegu** [comment]
@halseth , I'm not sure how to circumvent the coverage decrease. I added a test function to test the new code, and it increased the coverage at that package, but overall, in totally unrelated files, the coverage somehow decreased. I run the coverage test twice (updating the pull request) with the same results. Thanks!

**Comment by Bluetegu** [comment]
I did notice something strange when comparing the two coveralls coverage reports, run for the same code (builds 6014 and 6015). The report has a 'Jobs' tab, which has two lines, one summarizing the report result for the Go 1.94 job and one for the Go 1.10 job. In this report, build 6014, the two lines show (same) decrease in coverage, as should. In the previous run, build 6015, the Go 1.94 job shows increase in coverage (!) while the Go 1.10 line shows decrease in coverage (I have a screen capture of this part of the report if needed). So there might be something wrong with the coverage test or the way it is set up.

**Comment by Roasbeef** [comment]
Yeah the code coverage that we have on atm within the CI isn't very accurate as it isn't Go specific. I've even seen it _decrease_ code coverage if additional _comments_ were added lol.

**Comment by Bluetegu** [comment]
Its a long shot, but maybe its how coveralls is setup to handle parallel jobs? Coveralls response to this [issue](https://github.com/lemurheavy/coveralls-public/issues/1182) is to configure parallel processing using [this](http://docs.coveralls.io/parallel-build-webhook). Probably its already setup this way, but maybe it's worth double checking.

**Comment by Bluetegu** [comment]
I added a test function (testWipe at db_test.go) that helped cross the coverage threshold. The test is not really related to the logic of this pull request, but its helpful anyhow.

The parallel flag may still be an issue, and may explain the inconsistency in 'jobs' reported above, but it wasn't the culprit in the coverage calculation. I got the same results when doing the coverage locally.

File: rpcserver.go [hunk 5]
```diff
@@ -3152,14 +3154,36 @@ func (r *rpcServer) DescribeGraph(ctx context.Context,
...
-		edge := marshalDbEdge(edgeInfo, c1, c2)
-		resp.Edges = append(resp.Edges, edge)
+		// Do not include private channels unless specifically
+		// requested.
+		// A private channel does not need an authentication proof,
+		// while an advertized channel must have one. However in some
+		// situations, and specificatlly in integration tests (e.g. in
+		// open channel reorg test), public channels do not yet have
+		// an auth proof, hence relying on authProof != nil to determine
+		// if a channel is public would fail. Instead, to determine if the
```
Comment (halseth) [inline rpcserver.go]: Feels like this should be handled by the test instead?

If we rely on `authProof` here, the integration test should be aware that the channel needs 6 confirmations before it is public.

File: lnrpc/rpc.proto [hunk 6]
```diff
@@ -1498,6 +1498,11 @@ message ChannelEdge {
 }
 
 message ChannelGraphRequest {
+     /**
```
Comment (halseth) [inline lnrpc/rpc.proto]: All proto changes should be in its own commit :)

File: lnd_test.go [hunk 7]
```diff
@@ -4432,8 +4432,8 @@ func testPrivateChannels(net *lntest.NetworkHarness, t *harnessTest) {
 	// nodes know about. Carol and Alice should know about 4, while
 	// Bob and Dave should only know about 3, since one channel is
 	// private.
-	numChannels := func(node *lntest.HarnessNode) int {
-		req := &lnrpc.ChannelGraphRequest{}
+	numChannels := func(node *lntest.HarnessNode, private bool) int {
```
Comment (halseth) [inline lnd_test.go]: Nice 👍

File: channeldb/db.go [hunk 8]
```diff
@@ -534,6 +535,38 @@ func fetchChannels(d *DB, pending, waitingClose bool) ([]*OpenChannel, error) {
 	return channels, nil
 }
 
+// FetchPrivateOpenChannelIDs will return a map with the IDs of all open
+// private channels as keys, which can be used for checking if an edge in
+// the graph is private or not.
+// IDs of both 'default' open channels and 'waiting to be closed' private
+// channels are returned. IDs of pending opened channels are not returned,
+// as those are not yet advertised to the graph.
+func (d *DB) FetchPrivateOpenChannelIDs() (map[uint64]bool, error) {
```
Comment (halseth) [inline channeldb/db.go]: Would be awesome if this could be added in its own commit, as it is independent from the RPC change. However, I'm still not convinced we need this method, and cannot just rely on the `authProof`.

File: lnrpc/rpc.proto (hunk 6 above)
Comment (Bluetegu) [reply lnrpc/rpc.proto]: Should I break it to the following commits?
I'm placing the proto changes first, as those are the ones that will most likely need to change upon rebase if master moves forward before the merge. is this the right consideration?

Mesage: "Add private property to ChannelGraphRequest rpc"
Files:  lnrpc/*

Message: "Add unit test for db Wipe"
Files:   channeldb/db_test.go

Message: "Add FetchPrivateOpenChannelIDs to easily check if an edge is private or public."
Files:  channeldb/db.go, channeld/channel_test.go

Message: "Add support for private property of ChannelGraphRequest in DescribeGraph rpc server method"  
Files: rpcserver.go, lnd_test.go

Message: " Add --private flag to describegraph lncli command.
    
    Private channels are returned only if the private flag is set.
    Note: this changes the default output of describegraph which
    used to return all channels known to node."
Files: cmd/lncli/commands.go

Thanks!

File: rpcserver.go (hunk 5 above)
Comment (halseth) [reply rpcserver.go]: I guess it boils down to taste, what you feel is useful information given by the RPC call. 

To me, it sounds useful to let `describegraph` be the API for getting your nodes' "view of the network", and then checking the `authProof` would give a nice overview of which channels your node is considering "announced" (come to think of it, I don't think there exist any way currently to get this information). As mentioned, `ListChannels` can be used to get the "is private now, and staying private" information about your own channels. 

However, you have well thought of arguments, and I'm certainly not saying my opinion is the right one in this case. We can let others way in on the discussion if we don't reach an agreement 😄

File: lnrpc/rpc.proto (hunk 6 above)
Comment (halseth) [reply lnrpc/rpc.proto]: Yes! Sounds like a perfect split! 👍

File: rpcserver.go (hunk 5 above)
Comment (Bluetegu) [reply rpcserver.go]: I think the right approach is to add a separate 'announced' flag which if set would return only public channels that were announced. I feel its wrong to use 'private' flag to something that is not exactly 'private'; a channel shouldn't change from being private to being public, rather a public channel state changes from not having a confirmed auth proof to having one.
I can easily add an 'announced' flag too, if you think this makes sense.

Comment (Bluetegu) [reply rpcserver.go]: Or maybe add an 'announced' bool property to the edges in the describegraph response, so you got the info there and can further filter if needed.

Comment (halseth) [reply rpcserver.go]: What do you think about changing the flag's name to `included_not_announced` or similar?

Comment (halseth) [reply rpcserver.go]: Reason being there's a similar concern (not as big as for private channels though) to people using `describegraph` and by accident leak channels that are not yet announced, as they shouldn't be used before they are properly announced.

Comment (Bluetegu) [reply rpcserver.go]: Yes, this sounds right to me (--include_unannounced)

In terms of integration tests:
- I'll add the flag to Alice's describeGraph query in the reorg test, and add a comment there something along: "Request all edges as channel is opened but authentication proof has not been confirmed yet"
- In the private channels test I'll leave the test as in this pull request, just change the flag name.
- I'll add a new integration test that tests the new flag.

**Comment by Bluetegu** [comment]
Changes made. Ready for inspection :-)

File: lnrpc/rpc.proto [hunk 9]
```diff
@@ -1498,6 +1498,13 @@ message ChannelEdge {
...
+     Whether unannounced channels should be returned or not. If set,
```
Comment (halseth) [inline lnrpc/rpc.proto]: suggestions for text: ~~returned~~ included in the response

File: lnrpc/rpc.proto [hunk 10]
```diff
@@ -1498,6 +1498,13 @@ message ChannelEdge {
...
+     unannounced channels are returned.
+     Unannoucned channels are both private channels and public
```
Comment (halseth) [inline lnrpc/rpc.proto]: typo: Unannoucned

File: lnrpc/rpc.proto [hunk 11]
```diff
@@ -1498,6 +1498,13 @@ message ChannelEdge {
...
+     channels whose authentication proof has not been confirmed yet.
```
Comment (halseth) [inline lnrpc/rpc.proto]: suggestion: ~~whose authentication proof has not been confirmed yet~~ that are not yet announced to the network

File: channeldb/db_test.go [hunk 12]
```diff
@@ -33,3 +33,37 @@ func TestOpenWithCreate(t *testing.T) {
 		t.Fatalf("channeldb failed to create data directory")
 	}
 }
+
+func TestWipe(t *testing.T) {
```
Comment (halseth) [inline channeldb/db_test.go]: All new tests should have a godoc explaining what is being tested.

File: channeldb/db_test.go [hunk 13]
```diff
@@ -33,3 +33,37 @@ func TestOpenWithCreate(t *testing.T) {
...
+	cdb, err := Open(dbPath)
+	if err != nil {
+		t.Fatalf("unable to create channeldb: %v", err)
+	}
+	defer cdb.Close()
+
+	if err := cdb.Wipe(); err != nil {
+		t.Fatalf("unable to wipe channeldb: %v", err)
+	}
+	// check errors are returned
```
Comment (halseth) [inline channeldb/db_test.go]: Comments should be full sentences.

File: channeldb/db_test.go [hunk 14]
```diff
@@ -33,3 +33,37 @@ func TestOpenWithCreate(t *testing.T) {
...
+	// increase cover
+	_, err = cdb.FetchAllOpenChannels()
+	if err != ErrNoActiveChannels {
+		t.Fatalf("fetching open channels and expecting error: '%v' got: '%v'", ErrNoActiveChannels, err)
```
Comment (halseth) [inline channeldb/db_test.go]: lines here and below are too long. Should be wrapped at 80 chars with 8 char tab width.

File: lnd_test.go [hunk 15]
```diff
@@ -870,7 +870,7 @@ func assertChannelPolicy(t *harnessTest, node *lntest.HarnessNode,
 	advertisingNode string, expectedPolicy *lnrpc.RoutingPolicy,
 	chanPoints ...*lnrpc.ChannelPoint) {
 
-	descReq := &lnrpc.ChannelGraphRequest{}
+	descReq := &lnrpc.ChannelGraphRequest{IncludeUnannounced: true}
```
Comment (halseth) [inline lnd_test.go]: style nit:
```go
descReq := &lnrpc.ChannelGraphRequest{
         IncludeUnannounced: true,
}
```

File: lnd_test.go [hunk 16]
```diff
@@ -4122,6 +4122,85 @@ func testSendToRouteErrorPropagation(net *lntest.NetworkHarness, t *harnessTest)
 	closeChannelAndAssert(ctxt, t, net, carol, chanPointCarol, false)
 }
 
+// check unannounced channels are not returned by describeGraph RPC request
```
Comment (halseth) [inline lnd_test.go]: godoc should start "testUnannouncedChannels checks that...."

File: lnd_test.go [hunk 17]
```diff
@@ -4122,6 +4122,85 @@ func testSendToRouteErrorPropagation(net *lntest.NetworkHarness, t *harnessTest)
...
+	if err != nil {
+		t.Fatalf("unable to query alice's graph: %v", err)
+	}
+
+	numEdges := len(chanGraph.Edges)
+	if numEdges != 1 {
+		t.Fatalf("expected to find 1 edge in the graph, found %d", numEdges)
+	}
+
+	// But none already announced
```
Comment (halseth) [inline lnd_test.go]: nit: period at end.
Comments should be full sentences.

File: lnd_test.go [hunk 18]
```diff
@@ -4122,6 +4122,85 @@ func testSendToRouteErrorPropagation(net *lntest.NetworkHarness, t *harnessTest)
...
+	)
+	if err != nil {
+		t.Fatalf("unable to open channel: %v", err)
+	}
+
+	// Mine 2 blocks, and check that the channel is opened but not yet
+	// announced to the network
+	mineBlocks(t, net, 2)
+
+	// figure out the chanPoint, required to close the channel
```
Comment (halseth) [inline lnd_test.go]: use full sentences for comments.

File: lnd_test.go [hunk 19]
```diff
@@ -4122,6 +4122,85 @@ func testSendToRouteErrorPropagation(net *lntest.NetworkHarness, t *harnessTest)
...
+	if err != nil {
+		t.Fatalf("unable to query alice's graph: %v", err)
+	}
+
+	numEdges = len(chanGraph.Edges)
+	if numEdges != 1 {
+		t.Fatalf("expected to find 1 announced edge in the graph, found %d", numEdges)
+	}
+
+	// Close channel
```
Comment (halseth) [inline lnd_test.go]: full sentence.

File: lnd_test.go [hunk 20]
```diff
@@ -4122,6 +4122,85 @@ func testSendToRouteErrorPropagation(net *lntest.NetworkHarness, t *harnessTest)
...
+	numEdges = len(chanGraph.Edges)
+	if numEdges != 0 {
+		t.Fatalf("expected to find 0 announced edges in the graph, found %d", numEdges)
+	}
+
+	// Mine 4 more blocks, and check that the channel is now advertized
+	mineBlocks(t, net, 4)
+
+	// Give the network a chance to learn that auth proof is confirmed
+	time.Sleep(time.Millisecond * 50)
```
Comment (halseth) [inline lnd_test.go]: Instead of sleeping here, you should be using a `WaitPredicate`. Check other tests for how this is done :)

File: lnd_test.go [hunk 21]
```diff
@@ -4444,25 +4523,37 @@ func testPrivateChannels(net *lntest.NetworkHarness, t *harnessTest) {
...
 		}
-		carolChans := numChannels(carol)
+		carolChans := numChannels(carol, true)
 		if carolChans != 4 {
 			predErr = fmt.Errorf("expected Carol to know 4 edges, "+
 				"had %v", carolChans)
 			return false
 		}
-		daveChans := numChannels(dave)
+		carolPubChans := numChannels(carol, false)
```
Comment (halseth) [inline lnd_test.go]: Nice test addition! 👍

File: rpcserver.go [hunk 22]
```diff
@@ -3158,8 +3159,16 @@ func (r *rpcServer) DescribeGraph(ctx context.Context,
 	err = graph.ForEachChannel(func(edgeInfo *channeldb.ChannelEdgeInfo,
 		c1, c2 *channeldb.ChannelEdgePolicy) error {
 
-		edge := marshalDbEdge(edgeInfo, c1, c2)
-		resp.Edges = append(resp.Edges, edge)
+		// Do not include unannounced channels unless specifically
+		// requested.
```
Comment (halseth) [inline rpcserver.go]: nit: no need for new line.

File: rpcserver.go [hunk 23]
```diff
@@ -3158,8 +3159,16 @@ func (r *rpcServer) DescribeGraph(ctx context.Context,
...
+		// Unannounced channels include both private channels as well as
+		// public channels whose authentication proof were not confirmed
```
Comment (halseth) [inline rpcserver.go]: Good comment!

File: rpcserver.go [hunk 24]
```diff
@@ -3158,8 +3159,16 @@ func (r *rpcServer) DescribeGraph(ctx context.Context,
...
+		// yet, hence were not advertized.
+		if includeUnannounced || edgeInfo.AuthProof != nil {
```
Comment (halseth) [inline rpcserver.go]: style nit: To avoid the indentation you can do:
```go
if !includeUnannounced && edgeInfo.AuthProof == nil {
      return nil
}
....append edge...
return nil
```

Comment (Bluetegu) [reply rpcserver.go]: I missed this one, sorry. Why do you think its better to avoid the indentation?

**Comment by Bluetegu** [comment]
Thanks for the thorough review. I believe I fixed all of the issues, spare the last. I actually somehow missed it while doing the mods. Let me know if you want it done too, and if you find any other issue. Thanks!

File: rpcserver.go (hunk 24 above)
Comment (halseth) [reply rpcserver.go]: I generally like to use "exit early" in case of non-happy path. It's just a style preference though, so if you prefer doing it like this, I don't mind.

In this case it would also have made the diff smaller, which is also a plus ;)

Comment (halseth) [reply rpcserver.go]: On indentation: here it is not that bad, but imagine we had nested conditions we wanted to be satisfied before adding the edge:
```go
if condition1 {
      calculation....
      if condtiondion2 {
            ....more caluclation....
            if condition3 {
                   edges = append(edges, edge)
            }
      }
}
return nil
```

while we could be doing 
```go
if !condition1 {
     return nil
}

calculation....
if !condtiondion2 {
     return nil
}

....more caluclation....
if !condition3 {
    return nil
}
edges = append(edges, edge)
return nil
```

Comment (Bluetegu) [reply rpcserver.go]: Sounds reasonable. The condition is also a bit more clear this way.

**Comment by Bluetegu** [comment]
Rebased and took the opportunity to fix the last nit.

File: lnd_test.go (hunk 19 above)
Comment (halseth) [reply lnd_test.go]: ☝️

File: cmd/lncli/commands.go [hunk 25]
```diff
@@ -2467,15 +2467,26 @@ var describeGraphCommand = cli.Command{
 	Category: "Peers",
 	Description: "Prints a human readable version of the known channel " +
 		"graph from the PoV of the node",
-	Usage:  "Describe the network graph.",
+	Usage: "Describe the network graph.",
+	Flags: []cli.Flag{
+		cli.BoolFlag{
+			Name: "include_unannounced",
+			Usage: "If set, unannounced channels will be included in the graph. " +
+				"Unannounced channels include both private channels, and public " +
```
Comment (halseth) [inline cmd/lncli/commands.go]: Update comment to reflect the change you did to the proto comment

**Comment by Bluetegu** [comment]
I think I got them all :-)

**Comment by halseth** [comment]
There's still this one unresolved comment about making comment full sentence.

**Comment by Bluetegu** [comment]
I'm not a native English speaker so please bear with me on this.

The current comment is "Close channel"
When you say a full sentence you mean something like:
"Close the channel."
This is similar to the "Close all channels." comment on another test.
Or are you referring to something more than that?

BTW, do you want me to switch to closeChannelAndAssert?

**Comment by halseth** [comment]
No worries, sorry I didn't explain. "Close the channel." is a full sentence, even better would be "Close the channel used during the test.".

And yes, `closeChannelAndAssert` is preferred, good catch!

**Comment by Bluetegu** [comment]
Changes made. 

BTW, I'm not sure how the 'resolve conversation' here in the discussion of the pull request is supposed to work: do I need to press the 'resolve conversation' once I fix the issue, or is the reviewer (you) supposed to do that once the issue is indeed fixed, or we should simply ignore this github feature altogether.
